Add documentation to Random module

adlerhsieh opened the pull request.

jhass commented on src/random.cr:
```diff
@@ -1,35 +1,85 @@
 require "random/mt19937"
 
+# `Random` provides an interface for generating random values with `MT19937` library.
```
Using a pseudo random number generator (PRNG).

jhass commented on src/random.cr:
````diff
@@ -1,35 +1,85 @@
+# r.next_int  # => 2223112
+# ```
+#
+# It also defines global methods.
````
\`rand\`, \`Array#sample\` and \`Array#shuffle\` delegate here.

jhass commented on src/random.cr:
````diff
@@ -1,35 +1,85 @@
+# Random.rand(5)  # => 2
+# ```
+#
+# Generates values with an instance.
````
The above methods delegate to a shared instance, you can create your own too:

jhass commented on src/random.cr:
```diff
@@ -1,35 +1,85 @@
 module Random
   DEFAULT = MT19937.new
 
+  # Generates a random `UInt32`.
```
That's wrong, check again what the method does.

jhass commented on src/random.cr:
```diff
@@ -1,35 +1,85 @@
     Intrinsics.read_cycle_counter.to_u32
   end
 
+  # Initiates an instance with the given *seed*.
```
Mention the default?

jhass commented on src/random.cr:
```diff
@@ -1,35 +1,85 @@
     next_u32 * (1.0/4294967295.0)
   end
 
+  # Generates a random `Float64`.
```
Mention the range.

jhass commented on src/random.cr:
```diff
@@ -1,35 +1,85 @@
     next_float
   end
 
+  # Returns a random `Int32` between 0 and *x*.
```
Rename `x` to something more meaningful like `max`?

kirbyfan64 replied:
Also, maybe the range should explicitly be specified as inclusive or exclusive?

jhass commented on src/random.cr:
```diff
@@ -38,6 +88,11 @@ module Random
     end
   end
 
+  # Returns a random `Int32` in the given range *x*.
```
Rename `x`?

adlerhsieh commented:
Updated. I found several uses of `rand` in [Array#sample](https://github.com/adlerhsieh/crystal/blob/master/src/array.cr#L1333) but did not see any in `Array#shuffle`, maybe I missed something?

kirbyfan64 commented on src/random.cr:
```diff
@@ -1,66 +1,125 @@
-  def rand(x : Int) : Int32
-    if x > 0
-      (next_u32 % x).to_i32
+  # Returns a random `Int32` between 0 and *max*.
```
You should put "greater than 0 and less than *max*" to emphasize the value will always be less than the max.

kirbyfan64 commented on src/random.cr:
```diff
@@ -1,66 +1,125 @@
-  def rand(x : Range(Int, Int)) : Int32
-    span = x.end - x.begin
-    span += 1 unless x.excludes_end?
+  # Returns a random `Int32` in the given *range* (inclusive).
```
The "inclusive" isn't needed here (did I put my last line note on the wrong line?).

adlerhsieh replied:
Nope, I misplaced it.

kirbyfan64 commented on src/random.cr:
````diff
@@ -1,66 +1,125 @@
+# r.next_int  # => 2223112
+# ```
+#
+# It also defines a global method `#rand`, which `Array#sample` delegates.
````
Now it sounds like "It" is referring to the shared method implementation. What about "The module also defines a ..."?

adlerhsieh replied:
Yeah that sounds better.

jhass commented on src/random.cr:
````diff
@@ -1,66 +1,125 @@
+# Random.rand(5)  # => 2
+# ```
+#
+# The above methods delegate to a shared instance. You can create one like:
````
That sounds like it would create a new *shared* instance, which is not quite right.

kirbyfan64 commented on src/random.cr:
````diff
@@ -1,66 +1,125 @@
+# r.next_int  # => 2223112
+# ```
+#
+# The modules also defines a global method `#rand`, which `Array#sample` delegates.
````
Typo (modules => module). :)

jhass commented:
`Array#shuffle` delegates to `Pointer#shuffle!`
* https://github.com/manastech/crystal/blob/master/src/array.cr#L1406 
* https://github.com/manastech/crystal/blob/master/src/pointer.cr#L340

asterite commented on src/random.cr:
```diff
@@ -1,66 +1,126 @@
 module Random
   DEFAULT = MT19937.new
 
+  # Generates a `UInt32` based on current time using `read_cycle_counter` from LLVM.
```
I wouldn't mention this implementation detail

jhass replied:
I think how the seed is obtained is quite relevant to a PRNG. That it's an LLVM intrinsic is indeed irrelevant though, instead some short pointer as to what the "cycle counter" is would be good.

ysbaddaden replied:
We should make it clear that this is not a real and secure random number, unlike `SecureRandom.bytes`.

adlerhsieh commented:
Updated. I changed it to 

```ruby
# Generates a `UInt32` based on current timestamp using cycle counter.
def self.new_seed : UInt32
```

I think this is enough to cover where it comes from and its security level, and is not too detailed about implementation. What do you think about it?

jhass commented:
I think shortening it to "timestamp" is misleading, it's not epoch or something comparable.

> The ‘llvm.readcyclecounter‘ intrinsic provides access to the cycle counter register (or similar low latency, high accuracy clocks) on those targets that support it. On X86, it should map to RDTSC. On Alpha, it should map to RPCC. As the backing counters overflow quickly (on the order of 9 seconds on alpha), this should only be used for small timings.

http://llvm.org/docs/LangRef.html#llvm-readcyclecounter-intrinsic

> The Time Stamp Counter (TSC) is a 64-bit register present on all x86 processors since the Pentium. It counts the number of cycles since reset.

https://en.wikipedia.org/wiki/Time_Stamp_Counter

adlerhsieh commented:
Thanks for clarification. I specify it a bit further instead.

```crystal
# Converts to a `UInt32` from an accurate current time value 
# using the cycle counter register.
def self.new_seed : UInt32
```

jhass commented:
I think `time` is still misleading. It's not a point in real world time, it's simply a monotonic counter since the CPU was started.

adlerhsieh commented:
I see. How about:

```crystal
# Returns a `UInt32` converted from a counter value 
# generated by the cycle counter register.
def self.new_seed : UInt32
```

That would make more sense about the counter.

jhass commented:
I could live with  that, "read from" instead of "converted" perhaps. 

What do others think?

jhass commented:
Well, thanks then.

adlerhsieh commented:
Thanks!
